### PR TITLE
Muon analysis setting active groups on file load and disabling deselecting all groups/pairs

### LINKED ISSUE #29108: Default group/pair not being properly selected on Grouping table load in Muon Analysis
In the Muon Analysis GUI when you load in a grouping table no group/pair is selected to analyse. Instead the behavior should be:
 * If a default is specified this should be selected.
 * Otherwise if any pairs all pairs should be selected.
 * Otherwise all groups should be selected.

It should also not be possible to have no group/pair selected so de-selecting the last group/pair should be disabled.

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
**Description of work.**
Adds in default selection of groups and pairs when loading a grouping file from xml. Also checks whether a group or pair is the last selected when deselected and disallows the action if so.

**To test:**
 * Load some data into the Muon Analysis GUI.
 * In the grouping tab make a selection of groups and pairs and create a grouping file by clicking save
 * Close the GUI then re-open and re-load the data.
 * Load the new grouping table and check that the rules for default selection described in the issue are followed.
 * Try to unselect all groups and pairs and check deselecting the final group/pair is disabled.

Fixes #29108  

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: docs/source/release/v5.1.0/muon.rst [hunk 1]
```diff
@@ -47,6 +47,7 @@ Improvements
 - Tf asymmetry mode can now be performed on double pulse fits from the muon analysis GUI.
 - Updated :ref:`LoadMuonNexusV2 <algm-LoadMuonNexusV2>` and  :ref:`LoadPSIMuonBin <algm-LoadPSIMuonBin>` to load a list of time zeros into a new property TimZeroList.
 - Updated :ref:`LoadMuonNexusV2 <algm-LoadMuonNexusV2>` and  :ref:`LoadPSIMuonBin <algm-LoadPSIMuonBin>` to add an option to not auto-correct the time by loaded timezero.
+- Added in default group and pair selection when loading grouping files from xml.
```
Comment (AnthonyLim23) [inline docs/source/release/v5.1.0/muon.rst:50]: Is this more of a bug fix? I would have thought that on loading a grouping file something should be selected.

File: scripts/Muon/GUI/Common/grouping_tab_widget/grouping_tab_widget_presenter.py [hunk 2]
```diff
@@ -154,10 +154,16 @@ def handle_load_grouping_from_file(self):
             if pair.forward_group in self._model.group_names and pair.backward_group in self._model.group_names:
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/grouping_tab_widget/grouping_tab_widget_presenter.py:154]: should this have a try statement too?

File: scripts/Muon/GUI/Common/grouping_tab_widget/grouping_tab_widget_presenter.py [hunk 3]
```diff
@@ -154,10 +154,16 @@ def handle_load_grouping_from_file(self):
...
                 self._model.add_pair(pair)
 
+        if default:
+            if default in self._model.group_names:
+                self._model.add_group_to_analysis(default)
+            elif default in self._model.pair_names:
+                self._model.add_pair_to_analysis(default)
+
```
Comment (tolu28-coder) [inline scripts/Muon/GUI/Common/grouping_tab_widget/grouping_tab_widget_presenter.py:162]: It shouldn't happen but if the file doesn't define defaults then there wont be any groups or pairs selected so maybe there should be an else block to catch it if it occurs at all.

File: scripts/test/Muon/grouping_tab/grouping_tab_presenter_test.py [hunk 4]
```diff
@@ -166,6 +166,66 @@ def test_loading_does_not_insert_invalid_groups(self):
             self.assertEqual(self.grouping_table_view.num_rows(), 1)
             self.assertEqual(self.pairing_table_view.num_rows(), 0)
 
+    def test_loading_selects_all_pairs_if_any_pairs_exist_and_no_default_set(self):
+        self.view.show_file_browser_and_return_selection = mock.Mock(return_value="grouping.xml")
```
Comment (AnthonyLim23) [inline scripts/test/Muon/grouping_tab/grouping_tab_presenter_test.py:170]: could we have a helper function to create the groups/pairs? We could even make a function that does the block in the patch (give some of the arguments for the return value).

File: scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py [hunk 5]
```diff
@@ -370,6 +369,26 @@ def test_updating_range_min_to_be_greater_than_range_max_displays_warning_and_vi
         self.view.warning_popup.assert_called_with('Maximum of group asymmetry range must be greater than minimum')
         self.assertEqual(self.view.warning_popup.call_count, 2)
 
+    def test_group_cannot_be_unselected_if_is_last_group_or_pair_selected(self):
+        self.add_three_groups_to_table()
+        self.model.add_group_to_analysis("my_group_0")
+        self.presenter.selected_group_changed_notifier.notify_subscribers = mock.MagicMock()
+
```
Comment (AnthonyLim23) [inline scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py:376]: we should check we have just one group selected before we try to unselect it

File: scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py [hunk 6]
```diff
@@ -370,6 +369,26 @@ def test_updating_range_min_to_be_greater_than_range_max_displays_warning_and_vi
...
+        self.assertEqual(self.model.selected_groups, ["my_group_0"])
+        self.presenter.selected_group_changed_notifier.notify_subscribers.assert_not_called()
+
+    def test_last_group_to_analyse_can_be_removed_if_a_pair_is_selected(self):
+        self.add_three_groups_to_table()
+        self.model.add_group_to_analysis("my_group_0")
+        self.context._group_pair_context._selected_pairs.append('pair1')
+        self.presenter.selected_group_changed_notifier.notify_subscribers = mock.MagicMock()
+
```
Comment (AnthonyLim23) [inline scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py:387]: can we check that the pair is selected?

File: scripts/test/Muon/grouping_tab/pairing_table_presenter_test.py [hunk 7]
```diff
@@ -258,6 +258,29 @@ def test_that_adding_new_pair_creates_incremented_default_name(self):
         self.assertEqual(str(self.view.get_table_item_text(2, 0)), "pair_3")
         self.assertCountEqual(self.model.pair_names, ["pair_1", "pair_2", "pair_3"])
 
+    def test_that_pair_cannot_be_unselected_if_is_last_pair_or_group_selected(self):
+        self.add_two_pairs_to_table()
+        self.model.add_pair_to_analysis("my_pair_0")
+
+        self.presenter.selected_pair_changed_notifier.notify_subscribers = mock.MagicMock()
```
Comment (AnthonyLim23) [inline scripts/test/Muon/grouping_tab/pairing_table_presenter_test.py:265]: same as above

**Comment by AnthonyLim23** [comment]
If you clear the grouping table and then add a group, nothing is selected. We should make sure that the first group that is added is also selected

**Comment by Mantid-Matthew** [comment]
> If you clear the grouping table and then add a group, nothing is selected. We should make sure that the first group that is added is also selected

Do we even want to let the grouping table be cleared? As clearing it leaves you in the same situation of manually removing all groups and pairs which we are trying to avoid here.

**Comment by AnthonyLim23** [comment]
> > If you clear the grouping table and then add a group, nothing is selected. We should make sure that the first group that is added is also selected
> 
> Do we even want to let the grouping table be cleared? As clearing it leaves you in the same situation of manually removing all groups and pairs which we are trying to avoid here.

That is true. We could remove the clear button.

File: scripts/Muon/GUI/Common/grouping_tab_widget/grouping_tab_widget_presenter.py (hunk 3 above)
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/grouping_tab_widget/grouping_tab_widget_presenter.py:162]: The case of a file not having a default it actually relatively common. In this case the call to plot_default_groups_or_pairs below sets the selected groups/pars to there default instead.

Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/grouping_tab_widget/grouping_tab_widget_presenter.py:162]: Although this is not very clear from the code I will see if I can make it a bit less opaque.

File: scripts/test/Muon/grouping_tab/grouping_tab_presenter_test.py (hunk 4 above)
Comment (Mantid-Matthew) [reply scripts/test/Muon/grouping_tab/grouping_tab_presenter_test.py:170]: The setup of the groups/pairs is unique in each case here as they describe the outcome of the loading the grouping file so I don't think there is much to be gained from a helper function. I can pull the patch statement out though.

**Comment by Mantid-Matthew** [comment]
> > > If you clear the grouping table and then add a group, nothing is selected. We should make sure that the first group that is added is also selected
> > 
> > 
> > Do we even want to let the grouping table be cleared? As clearing it leaves you in the same situation of manually removing all groups and pairs which we are trying to avoid here.
> 
> That is true. We could remove the clear button.

Yes I think we should either allow all groups and pairs to be removed and have the GUI handle this gracefully or remove the clear button.  I think the nicer solution is the former as I can envisage workflows where you want to start from a blank slate. If we go with this however 'd suggest not doing it as part of this PR.

**Comment by AnthonyLim23** [comment]
Would it be much work to get it to handle it gracefully?

**Comment by AnthonyLim23** [comment]
I have found a way to get in an odd state. Have one group/pair selected (but other groups/pairs are present). Then delete the selected group/pair, we have nothing selected. I think this might be too niche to fix.

**Comment by Mantid-Matthew** [comment]
> I have found a way to get in an odd state. Have one group/pair selected (but other groups/pairs are present). Then delete the selected group/pair, we have nothing selected. I think this might be too niche to fix.

Yes I think I would agree here. The major improvement here is selecting groups sensibly on file loading. Upon consideration I am slightly unconvinced that we need to make sure that the user cannot manually un-select everything at all, particularly as this can currently lead to a case where they can't un-select a group without knowing why.

**Comment by AnthonyLim23** [comment]
we prevent them deselecting everything because it can cause other bugs later on

**Comment by Mantid-Matthew** [comment]
Yes but I think I would argue the more intuitive solution here is to handle having nothing selected. Rather than eliminating every way to get to this state. But I think this PR is an improvement on the current functionality so we shouldn't hold up it going in too much.

**Comment by Mantid-Matthew** [comment]
If I had a look into why it breaks when nothing is selected and it turned out to be relatively easy to fix could I persuade you to go down this route instead? Apologies for changing my mind slightly on this, as I had to put more and more special cases in to handle nothing being selected it just seemed like the neater, less complex and more maintainable solution to make the state where nothing is selected stable.

**Comment by AnthonyLim23** [comment]
do it

**Comment by AnthonyLim23** [comment]
can we disable the fit buttons (fitting and sequential tabs) when no groups/pairs are selected? The error message at the moment is unhelpful (its about the function)

**Comment by Mantid-Matthew** [comment]
> can we disable the fit buttons (fitting and sequential tabs) when no groups/pairs are selected? The error message at the moment is unhelpful (its about the function)

Would improving the error message in this case be an acceptable solution?

**Comment by AnthonyLim23** [comment]
yes

**Comment by AnthonyLim23** [comment]
changing the instrument is causing a consistent crash for me on this branch. For some reason it says that python has stopped working

**Review by AnthonyLim23 — APPROVED** [review]

**Pull request merged**